### PR TITLE
[ASR] add amp for U2 conformer, test=asr

### PR DESCRIPTION
**Pull request by zxcd**
### PR types

New features
### PR changes

Models
### Describe

add amp for U2 conformer

### CODE REVIEW

File: paddlespeech/s2t/exps/u2/model.py [hunk 1]
```diff
@@ -47,14 +47,19 @@ class U2Trainer(Trainer):
...
 
         # forward
         utt, audio, audio_len, text, text_len = batch_data
-        loss, attention_loss, ctc_loss = self.model(audio, audio_len, text,
-                                                    text_len)
+        if scaler:
+            with paddle.amp.auto_cast(level=self.amp_level):
+                loss, attention_loss, ctc_loss = self.model(audio, audio_len,
+                                                            text, text_len)
+        else:
```
Comment (zh794390558) [inline paddlespeech/s2t/exps/u2/model.py]: 可以用enable来控制。

File: paddlespeech/s2t/exps/u2/model.py [hunk 2]
```diff
@@ -77,12 +82,19 @@ def train_batch(self, batch_index, batch_data, msg):
             # processes.
             context = nullcontext
         with context():
-            loss.backward()
+            if scaler:
+                scaled = scaler.scale(loss)
```
Comment (zh794390558) [inline paddlespeech/s2t/exps/u2/model.py]: scaler.scale(loss).backward()

File: paddlespeech/s2t/exps/u2/model.py [hunk 3]
```diff
@@ -77,12 +82,19 @@ def train_batch(self, batch_index, batch_data, msg):
...
+                scaled.backward()
+            else:
+                loss.backward()
             layer_tools.print_grads(self.model, print_func=None)
 
         # optimizer step
         if (batch_index + 1) % train_conf.accum_grad == 0:
-            self.optimizer.step()
+            if scaler:
+                scaler.minimize(self.optimizer, scaled)
```
Comment (zh794390558) [inline paddlespeech/s2t/exps/u2/model.py]: scaler.step()
scaler.update()
分开写吧。

File: paddlespeech/s2t/exps/u2/model.py [hunk 4]
```diff
@@ -77,12 +82,19 @@ def train_batch(self, batch_index, batch_data, msg):
...
+                scaled = scaler.scale(loss)
+                scaled.backward()
+            else:
+                loss.backward()
             layer_tools.print_grads(self.model, print_func=None)
 
         # optimizer step
         if (batch_index + 1) % train_conf.accum_grad == 0:
-            self.optimizer.step()
+            if scaler:
```
Comment (zh794390558) [inline paddlespeech/s2t/exps/u2/model.py:96]: grad_clip放在这里吧。

**Review by zh794390558 — COMMENTED** [review]
ckpt save/resume 没有加。

File: paddlespeech/s2t/exps/u2/model.py (hunk 4 above)
Comment (zxcd) [reply paddlespeech/s2t/exps/u2/model.py:96]: grad_clip在optimizer内部，不需要拆开写？

File: paddlespeech/s2t/exps/u2/model.py [hunk 5]
```diff
@@ -253,6 +266,18 @@ def setup_model(self):
                 model_conf.output_dim = self.test_loader.vocab_size
 
         model = U2Model.from_config(model_conf)
+
+        # For Mixed Precision Training
+        self.scaler = self.config.get("use_amp", True)
```
Comment (zh794390558) [inline paddlespeech/s2t/exps/u2/model.py]: self.use_amp

**Review by zh794390558 — COMMENTED** [review]
inference的时候可以加上auto_cast，不加scalar。

File: paddlespeech/s2t/exps/u2/model.py (hunk 4 above)
Comment (zh794390558) [reply paddlespeech/s2t/exps/u2/model.py:96]: > grad_clip在optimizer内部，不需要拆开写？

干掉。用新的API来做。

之后和opt一起的都干掉。
